### PR TITLE
[SM-150] Project / Secret mapping

### PR DESCRIPTION
**Pull request by cd-bitwarden**
## Type of change

```
- [ x ] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Fixing UI's handling of the secrets and projects list response and displaying them properly in the UI

## Code changes

- **file.ext:** Description of what was changed and why

## Screenshots

image

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: bitwarden_license/bit-web/src/app/secrets-manager/models/view/secret-list.view.ts [hunk 1]
```diff
@@ -1,9 +1,12 @@
...
 
+import { ProjectsMappedToSecretResponse } from "../../secrets/responses/projects-mapped-to-secret-response";
+
 export class SecretListView implements View {
   id: string;
   organizationId: string;
   name: string;
   creationDate: string;
   revisionDate: string;
+  projects: ProjectsMappedToSecretResponse[];
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/models/view/secret-list.view.ts:11]: This should be a decrypted view.

File: .vscode/launch.json [hunk 2]
```diff
@@ -1,31 +0,0 @@
-{
```
Comment (Hinton) [inline .vscode/launch.json:1]: Please don't delete this file.

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/projects-mapped-to-secret-response.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { BaseResponse } from "../../../../../../../libs/common/src/models/response/base.response";
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/projects-mapped-to-secret-response.ts:1]: Please use the `@bitwarden/common` path.

Comment (Hinton) [reply bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/projects-mapped-to-secret-response.ts:1]: The filename should be `projects-mapped-to-secret.response.ts`

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/secret-with-projects-list-response.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { BaseResponse } from "../../../../../../../libs/common/src/models/response/base.response";
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/secret-with-projects-list-response.ts:1]: Import path.

Comment (Hinton) [reply bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/secret-with-projects-list-response.ts:1]: Filename.

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/secrets-list.component.html [hunk 5]
```diff
@@ -57,8 +57,14 @@
       </td>
       <td bitCell>{{ secret.name }}</td>
       <td bitCell>
-        <!-- TODO dynamically add project badge associated with specific secret once projects are implemented-->
-        <span bitBadge badgeType="secondary"> web app </span>
+        <span
+          *ngFor="let projectSecretMapping of secret.projects"
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/secrets/secrets-list.component.html:61]:
```suggestion
          *ngFor="let project of secret.projects"
```

File: bitwarden_license/bit-web/src/app/secrets-manager/models/view/secret-list.view.ts (hunk 1 above)
Comment (cd-bitwarden) [reply bitwarden_license/bit-web/src/app/secrets-manager/models/view/secret-list.view.ts:11]: okay, the decryption happens in secret.service.ts now

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/secret.service.ts [hunk 6]
```diff
@@ -142,8 +152,30 @@ export class SecretService {
...
         return secretListView;
       })
     );
   }
+
+  private async decryptProjectsMappedToSecrets(
+    organizationId: string,
+    projects: ProjectsMappedToSecretResponse[]
+  ): Promise<ProjectsMappedToSecretListView[]> {
+    const orgKey = await this.getOrganizationKey(organizationId);
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/secrets/secret.service.ts:167]: We already have the org key on line 137.

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/secret.service.ts [hunk 7]
```diff
@@ -43,8 +45,9 @@ export class SecretService {
       true,
       true
     );
-    const results = new ListResponse(r, SecretListItemResponse);
-    return await this.createSecretsListView(organizationId, results.data);
+
+    const results = new SecretWithProjectsListResponse(r);
+    return await this.createSecretsListView(organizationId, results.secrets, results.projects);
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/secrets/secret.service.ts:50]: I think we can send the `SecretWithProjectsListResponse` directly which makes for a slightly cleaner api.

File: bitwarden_license/bit-web/src/app/secrets-manager/models/view/projects-mapped-to-secret-list.view.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import { View } from "@bitwarden/common/models/view/view";
+
+export class ProjectsMappedToSecretListView implements View {
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/models/view/projects-mapped-to-secret-list.view.ts:3]: This name is a bit verbose, what do you think of just calling it `SecretProjectView`?

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/projects-mapped-to-secret.response.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { BaseResponse } from "@bitwarden/common/models/response/base.response";
+
+export class ProjectsMappedToSecretResponse extends BaseResponse {
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/projects-mapped-to-secret.response.ts:3]: `SecretProjectResponse`?

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/secret-with-projects-list.response.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+export class SecretWithProjectsListResponse extends BaseResponse {
+  secrets: SecretListItemResponse[];
+  projects: ProjectsMappedToSecretResponse[];
+
+  constructor(response: any) {
+    super(response);
+    const secrets = this.getResponseProperty("secrets");
+    const projects = this.getResponseProperty("projects");
+    this.projects =
+      projects == null ? null : projects.map((k: any) => new ProjectsMappedToSecretResponse(k)); //decrypt the project names here?
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/secret-with-projects-list.response.ts:15]:
```suggestion
      projects == null ? null : projects.map((k: any) => new ProjectsMappedToSecretResponse(k));
```

Comment (Hinton) [reply bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/secret-with-projects-list.response.ts:15]: Could we just do `projects?.map((k: any) => new ProjectsMappedToSecretResponse(k))`?

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/secret.service.ts (hunk 6 above)
Comment (cd-bitwarden) [reply bitwarden_license/bit-web/src/app/secrets-manager/secrets/secret.service.ts:167]: oh you're so right. Fixed

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/secret.service.ts (hunk 7 above)
Comment (cd-bitwarden) [reply bitwarden_license/bit-web/src/app/secrets-manager/secrets/secret.service.ts:50]: Sounds like a plan

File: bitwarden_license/bit-web/src/app/secrets-manager/models/view/projects-mapped-to-secret-list.view.ts (hunk 8 above)
Comment (cd-bitwarden) [reply bitwarden_license/bit-web/src/app/secrets-manager/models/view/projects-mapped-to-secret-list.view.ts:3]: I like that change, done.

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/secret-with-projects-list.response.ts (hunk 10 above)
Comment (cd-bitwarden) [reply bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/secret-with-projects-list.response.ts:15]: yes, that worked when I tested it out.

File: bitwarden_license/bit-web/src/app/secrets-manager/models/view/secret-project-view.ts [hunk 11]
```diff
@@ -0,0 +1,6 @@
+import { View } from "@bitwarden/common/models/view/view";
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/models/view/secret-project-view.ts:1]: Filename is wrong, should be `secret-project.view.ts`.

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/secret.service.ts (hunk 7 above)
Comment (Hinton) [reply bitwarden_license/bit-web/src/app/secrets-manager/secrets/secret.service.ts]: I think you missunderstood, we still need to do the mapping. But we could have a single input in 

```
const results = new SecretWithProjectsListResponse(r);
return await this.createSecretsListView(organizationId, results);
```

File: bitwarden_license/bit-web/src/app/secrets-manager/models/view/secret-project-view.ts (hunk 11 above)
Comment (cd-bitwarden) [reply bitwarden_license/bit-web/src/app/secrets-manager/models/view/secret-project-view.ts:1]: oops

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/secret-with-projects-list.response.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+export class SecretWithProjectsListResponse extends BaseResponse {
+  secrets: SecretListItemResponse[];
+  projects: SecretProjectResponse[];
+
+  constructor(response: any) {
+    super(response);
+    const secrets = this.getResponseProperty("secrets");
+    const projects = this.getResponseProperty("projects");
+    this.projects =
+      projects == null ? null : projects.map((k: any) => new SecretProjectResponse(k)); //decrypt the project names here?
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/secrets/responses/secret-with-projects-list.response.ts:15]:
```suggestion
      projects == null ? null : projects.map((k: any) => new SecretProjectResponse(k));
```

**Review by Hinton — APPROVED** [review]

**Pull request merged**